src/shims/sync.rs

```diff
@@ -392,41 +405,39 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
                 {
                     this.eval_libc_i32("EBUSY")
                 } else if kind == this.eval_libc("PTHREAD_MUTEX_RECURSIVE")? {
-                    match locked_count.checked_add(1) {
-                        Some(new_count) => {
-                            mutex_set_locked_count(this, mutex_op, Scalar::from_u32(new_count))?;
-                            Ok(0)
-                        }
-                        None => this.eval_libc_i32("EAGAIN"),
-                    }
+                    this.mutex_lock(id, active_thread);
+                    Ok(0)
                 } else {
                     throw_ub_format!(
                         "called pthread_mutex_trylock on an unsupported type of mutex"
                     );
                 }
             }
+        } else {
+            // The mutex is unlocked. Let's lock it.
+            this.mutex_lock(id, active_thread);
+            Ok(0)
         }
     }
 
     fn pthread_mutex_unlock(&mut self, mutex_op: OpTy<'tcx, Tag>) -> InterpResult<'tcx, i32> {
         let this = self.eval_context_mut();
 
         let kind = mutex_get_kind(this, mutex_op)?.not_undef()?;
-        let locked_count = mutex_get_locked_count(this, mutex_op)?.to_u32()?;
-        let owner_thread: ThreadId = mutex_get_owner(this, mutex_op)?.to_u32()?.into();
-
-        if owner_thread != this.get_active_thread()? {
-            throw_ub_format!("called pthread_mutex_unlock on a mutex owned by another thread");
-        } else if locked_count == 1 {
-            let blockset = mutex_get_or_create_blockset(this, mutex_op)?;
-            if let Some(new_owner) = this.unblock_some_thread(blockset)? {
-                // We have at least one thread waiting on this mutex. Transfer
-                // ownership to it.
-                mutex_set_owner(this, mutex_op, new_owner.to_u32_scalar())?;
-            } else {
-                // No thread is waiting on this mutex.
-                mutex_set_owner(this, mutex_op, Scalar::from_u32(0))?;
-                mutex_set_locked_count(this, mutex_op, Scalar::from_u32(0))?;
+        let id = mutex_get_or_create_id(this, mutex_op)?;
+
+        if let Some((owner_thread, current_locked_count)) = this.mutex_unlock(id) {
+            if owner_thread != this.get_active_thread()? {
+                throw_ub_format!("called pthread_mutex_unlock on a mutex owned by another thread");
```

Comment:
According to [this](https://linux.die.net/man/3/pthread_mutex_lock), this is not UB for PTHREAD_MUTEX_RECURSIVE and  PTHREAD_MUTEX_ERRORCHECK, it should actually return an error (and not actually release the lock).

Probably a good idea to directly link to the docs here. :D

Reply:
Actually probably it is better to link to [the POSIX version](https://pubs.opengroup.org/onlinepubs/9699919799/functions/pthread_mutex_unlock.html) which is not Linux-specific.

I'd expect that to guarantee strictly less, but the Linux docs say

> If the mutex type is PTHREAD_MUTEX_DEFAULT, attempting to recursively lock the mutex results in undefined behavior. 

where the POSIX docs say

> If the mutex type is PTHREAD_MUTEX_DEFAULT, the behavior of pthread_mutex_lock() may correspond to one of the three other standard mutex types as described in the table above.

Reply:
Thinking about it... the POSIX docs also imply that in general, reentrancy with PTHREAD_MUTEX_DEFAULT is UB. So we shouldn't be able to rely on it not being UB. But is there any way we can actually make it UB in Miri? AFAIK PTHREAD_MUTEX_DEFAULT has the same value as PTHREAD_MUTEX_NORMAL and there is nothing Miri can do about that...

Reply:
So on platforms where `PTHREAD_MUTEX_DEFAULT == PTHREAD_MUTEX_NORMAL`, I guess behavior effectively *is* defined here.

So, "unlocked mutex owned by other thread" should go through the same code as "unlocked mutex not owned by anyone". And then I think the best way to make the PTHREAD_MUTEX_DEFAULT situation explicit is to have another `else if kind == this.eval_libc("PTHREAD_MUTEX_DEFAULT")?` branch that raises UB, with a comment that this is only reachable on platforms where `DEFAULT` does not equal one of the other types.

Reply:
Uh... except that [glibc actually violated that idea](https://github.com/rust-lang/rust/issues/33770#issuecomment-220847521) at some point, and `DEFAULT` (when not explicitly set) did *not* have the same behavior as `NORMAL`. So please leave a FIXME in `mutex_init`; I have some ideas for how we might check that but we do not have to do that in this PR.

Reply:
I am confused:

> OTOH, on platforms where `PTHREAD_MUTEX_DEFAULT == PTHREAD_MUTEX_NORMAL`, I guess behaviour effectively is defined here.

How have you deduced this? The POSIX [documentation](https://pubs.opengroup.org/onlinepubs/9699919799/functions/pthread_mutex_unlock.html) has a table that for `NORMAL` in column *Unlock When Not Owner* has **undefined behavior**. The Linux [documentation](https://linux.die.net/man/3/pthread_mutex_lock) says: *If the mutex type is PTHREAD_MUTEX_NORMAL, <...> If a thread attempts to unlock a mutex that it has not locked <...> **undefined behavior** results.*

Reply:
> How have you deduced this? 

I figured that explicitly setting the mutex kind to `PTHREAD_MUTEX_DEFAULT` would have to be a NOP. It's the default, after all. But since then I learned that glibc actually disagrees with this, which IMO is rather insane but what can you do.^^

But I see you opened https://github.com/rust-lang/miri/issues/1419, so let's continue there.

Reply:
It is still wrong to raise UB here for some mutex kinds, as noted in the first post in this thread. We can resolve the details later together with https://github.com/rust-lang/miri/issues/1419, but please leave a FIXME.